### PR TITLE
Basic Migrate: Creating tables And insertion of data.

### PR DESCRIPTION
**Pull request by ralmsdeveloper**
Basic implementation of table creation and data insertion was done. We can make progress if this is good for you!

### CODE REVIEW

**Comment by ralmsdeveloper** [comment]
@cincuranet  The SQL output looks like this example:

```sql
CREATE TABLE "Author" (
    "AuthorId" BIGINT GENERATED BY DEFAULT AS IDENTITY NOT NULL,
    "TestBytes" BLOB SUB_TYPE BINARY,
    "TestDate" TIMESTAMP NOT NULL ,
    "TestDecimal" DECIMAL(18,2) NOT NULL ,
    "TestDouble" DOUBLE PRECISION NOT NULL ,
    "TestGuid" CHAR(16) CHARACTER SET OCTETS NOT NULL ,
    "TestInt" INTEGER NOT NULL ,
    "TestString" VARCHAR(100),
    CONSTRAINT "PK_Author" PRIMARY KEY ("AuthorId")
);
 
CREATE TABLE "Person" (
    "Name" VARCHAR(100) NOT NULL ,
    "LastName" VARCHAR(100) NOT NULL ,
    CONSTRAINT "PK_Person" PRIMARY KEY ("Name", "LastName")
);
  
CREATE TABLE "Book" (
    "BookId" BIGINT NOT NULL ,
    "AuthorId" BIGINT NOT NULL ,
    "Title" VARCHAR(100),
    CONSTRAINT "PK_Book" PRIMARY KEY ("BookId"),
    CONSTRAINT "FK_Book_Author_AuthorId" FOREIGN KEY ("AuthorId") REFERENCES "Author" ("AuthorId") ON DELETE CASCADE
);

```

**Comment by cincuranet** [comment]
Can you fix your whitespaces?

**Comment by ralmsdeveloper** [comment]
Jiri, The tabulation I left with 4 spaces. Do you prefer tab instead of 4 spaces?

**Comment by cincuranet** [comment]
https://github.com/cincuranet/FirebirdSql.Data.FirebirdClient/blob/master/Provider/src/.editorconfig

**Comment by ralmsdeveloper** [comment]
Okay, corrected!

I will use this formatting now!

**Comment by cincuranet** [comment]
The file `FbUpdateSqlGenerator.cs` should not be modified.

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs [hunk 1]
```diff
@@ -28,15 +28,15 @@ public static class FbDbContextOptionsExtensions
 	{
 		public static DbContextOptionsBuilder UseFirebird(this DbContextOptionsBuilder optionsBuilder, string connectionString, Action<FbDbContextOptionsBuilder> fbOptionsAction = null)
 		{
-			var extension = GetOrCreateExtension(optionsBuilder).WithConnectionString(connectionString);
+			var extension = (FbOptionsExtension)GetOrCreateExtension(optionsBuilder).WithConnectionString(connectionString);
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs]: What's the reason for the casting?

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs [hunk 2]
```diff
@@ -28,15 +28,15 @@ public static class FbDbContextOptionsExtensions
...
 			((IDbContextOptionsBuilderInfrastructure)optionsBuilder).AddOrUpdateExtension(extension);
 			fbOptionsAction?.Invoke(new FbDbContextOptionsBuilder(optionsBuilder));
 			return optionsBuilder;
 		}
 
 		public static DbContextOptionsBuilder UseFirebird(this DbContextOptionsBuilder optionsBuilder, DbConnection connection, Action<FbDbContextOptionsBuilder> fbOptionsAction = null)
 		{
-			var extension = GetOrCreateExtension(optionsBuilder).WithConnection(connection);
+			var extension = (FbOptionsExtension)GetOrCreateExtension(optionsBuilder).WithConnection(connection);
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs]: What's the reason for the casting?

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Metadata/FbValueGenerationStrategy.cs [hunk 3]
```diff
@@ -18,8 +18,9 @@
 namespace FirebirdSql.EntityFrameworkCore.Firebird.Metadata
 {
 	public enum FbValueGenerationStrategy
-    {
+	{
 		SequenceTrigger,
 		IdentityColumn,
-    }
+		None
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Metadata/FbValueGenerationStrategy.cs]: `None` should be first.

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs (hunk 1 above)
Comment (ralmsdeveloper) [reply Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs]: You will not be able to recover the connection properties for example!

With this -> options.FindExtension<FbOptionsExtension>() in FbOptions


And I already tested !!!

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs (hunk 2 above)
Comment (ralmsdeveloper) [reply Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs]: You will not be able to recover the connection properties for example!

With this -> options.FindExtension() in FbOptions

And I already tested !!!

**Comment by cincuranet** [comment]
I don't understand your English. So better show me some code.

**Comment by ralmsdeveloper** [comment]
I do not think fluency should be the point here, but is it if you accept collaboration?

The code has been informed, we can improve ... just say what you need!

**Comment by cincuranet** [comment]
Unless you can answer my questions I can't move forward with it. It's
not about fluency. It's about explaining your changes.

If you don't want to spend time on that don't issue PRs. It's then waste
of time for you and me as well.

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs (hunk 1 above)
Comment (cincuranet) [reply Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs]: Ahh, now I get it. The email notification missed the images. The `WithConnection` is fucking the type.

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Migrations/FbMigrationsSqlGenerator.cs [hunk 4]
```diff
@@ -15,15 +15,161 @@
...
+			else
+				base.ForeignKeyAction(referentialAction, builder);
+		}
+
+		private FbValueGenerationStrategy IsSequenceIdentityOrTrigger(FbValueGenerationStrategy? valueGenerate)
+		{
+			return valueGenerate ?? FbValueGenerationStrategy.None;
+		}
+
+		private void GenerateSequenceOnTrigger(MigrationCommandListBuilder builder, string table, string column)
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Migrations/FbMigrationsSqlGenerator.cs]: I think this should use interface similar what's [here](https://github.com/cincuranet/FirebirdSql.Data.FirebirdClient/blob/master/Provider/src/EntityFramework.Firebird/DefaultFbMigrationSqlGeneratorBehavior.cs). This is often opinionated and should be easily replaceable.

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Update/Internal/FbUpdateSqlGenerator.cs [hunk 5]
```diff
@@ -1,16 +1,16 @@
 ﻿/*
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Update/Internal/FbUpdateSqlGenerator.cs]: The file `FbUpdateSqlGenerator.cs` should not be modified.

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Metadata/FbValueGenerationStrategy.cs [hunk 6]
```diff
@@ -18,8 +18,9 @@
 namespace FirebirdSql.EntityFrameworkCore.Firebird.Metadata
 {
 	public enum FbValueGenerationStrategy
-    {
+	{
+		None,
 		SequenceTrigger,
-		IdentityColumn,
-    }
+		IdentityColumn
```
Comment (cincuranet) [inline Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Metadata/FbValueGenerationStrategy.cs]: Missing trailing comma.

**Comment by ralmsdeveloper** [comment]
The intention is to help, I will try to improve my English, and my communication better!

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Metadata/FbValueGenerationStrategy.cs (hunk 6 above)
Comment (ralmsdeveloper) [reply Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Metadata/FbValueGenerationStrategy.cs]: In being the last identifier, I did not need this, but I'll put it now!

Thanks!

Comment (cincuranet) [reply Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Metadata/FbValueGenerationStrategy.cs]: I know. But I kind of like putting it there anyway. It helps future changes of the code.

**Comment by ralmsdeveloper** [comment]
@cincuranet, See if it has improved!

**Comment by cincuranet** [comment]
Well there're still pending issues. But rather than doing fifteen
thousand iterations, I'll merge it manually and fix those myself. It
would be faster.

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Metadata/FbValueGenerationStrategy.cs (hunk 3 above)
Comment (ralmsdeveloper) [reply Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Metadata/FbValueGenerationStrategy.cs]: Done!

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs (hunk 1 above)
Comment (ralmsdeveloper) [reply Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs]: As I explained, we can use this CAST!

File: Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs (hunk 2 above)
Comment (ralmsdeveloper) [reply Provider/src/FirebirdSql.EntityFrameworkCore.Firebird/Extensions/FbDbContextOptionsExtensions.cs]: As I explained, we can use this CAST!

**Comment by cincuranet** [comment]
I'm going to close this PR. As I'm merging it I realized there's quite a few bugs. I'll cherry pick some pieces, but rest I have to rewrite. It would blow immediately.

I'm finishing it in `migrations-create-table` branch.

**Comment by ralmsdeveloper** [comment]
Feel free!
Do what you think is best.

**Pull request closed without merging**